src/ascii/chain.rs

```diff
@@ -0,0 +1,386 @@
+//! Types for accessing individual devices/axes on a chain of devices.
+
+use crate::{
+    ascii::{
+        check,
+        marker::Markers,
+        setting::{AxisSettings, DeviceSettings, Settings},
+        Port, Target,
+    },
+    backend::Backend,
+    error::{AsciiCheckDataError, AsciiError, DuplicateAddressError},
+};
+use std::{
+    cell::RefCell,
+    num::NonZeroU8,
+    rc::Rc,
+    sync::{Arc, Mutex},
+};
+
+/// Represents a chain of devices.
+#[derive(Debug)]
+pub struct Chain<'a, B, P = Rc<RefCell<Port<'a, B>>>> {
+    port: P,
+    info: Vec<DeviceInfo>,
+    _markers: Markers<'a, B>,
+}
+
+impl<'a, B, P> Chain<'a, B, P> {
+    /// Get the number of devices in the chain.
+    pub fn len(&self) -> usize {
+        self.info.len()
+    }
+
+    /// Get whether the chain has no devices in it.
+    pub fn is_empty(&self) -> bool {
+        self.info.is_empty()
+    }
+
+    /// Get an iterator over the devices in the chain.
+    pub fn iter(&self) -> ChainIter<'_, 'a, B, P> {
+        ChainIter {
+            chain: self,
+            next_device_address: unsafe { NonZeroU8::new_unchecked(1) },
+        }
+    }
+}
+
+impl<'a, B, P: Clone> Chain<'a, B, P> {
+    /// Get the [`Device`] at the specified address.
+    ///
+    /// Returns `None` if there is no device at that address.
+    pub fn device(&self, address: u8) -> Option<Device<'a, B, P>> {
+        self.info
+            .iter()
+            .find(|info| info.address.get() == address)
+            .map(|info| Device {
+                port: P::clone(&self.port),
+                info: *info,
+                _markers: Markers::default(),
+            })
+    }
+
+    /// Get the shared [`Port`] used to communicate with all devices on the chain.
+    pub fn port(&self) -> P {
```

Comment:
This could cause `panic`s if users hold it while calling methods on an `Axis`/`Device`/`Setting`. A warning is warranted.

Reply:
This has been removed